File: ts-sdk/src/billing.ts
/**
 * Billing API - balance and transactions
 */
import type { HTTPClient } from './http.js';

export interface Balance {
  total: string;
  rewards: string;
  paid: string;
  available: string;
  pendingReservations?: string;
  currency?: string;
  decimals?: number;
  userId?: string;
}

export interface Transaction {
  id: string;
  userId: string;
  amount: number;
  amountUsd: string;
  transactionType: string;
  status: string;
  rewards: boolean;
  jobId: string | null;
  createdAt: string;
  updatedAt: string;
  expiresAt: string | null;
  meta: Record<string, any> | null;
}

export interface ListTransactionsOptions {
  page?: number;
  pageSize?: number;
  transactionType?: string;
  status?: string;
  jobId?: string;
}

export interface TransactionListResponse {
  transactions: Transaction[];
  totalCount: number;
}

export interface TopUpCheckoutSession {
  sessionId: string;
  checkoutUrl: string;
}

function balanceFromDict(data: any): Balance {
  return {
    total: data.total_balance || '0',
    rewards: data.rewards_balance || '0',
    paid: data.balance || '0',
    available: data.available_balance || '0',
    pendingReservations: data.pending_reservations || '0',
    currency: data.currency || 'USD',
    decimals: data.decimals ?? 2,
    userId: data.user_id || '',
  };
}

function transactionFromDict(data: any): Transaction {
  return {
    id: data.id || '',
    userId: data.user_id || '',
    amount: data.amount || 0,
    amountUsd: data.amount_usd || 0,
    transactionType: data.transaction_type || '',
    status: data.status || '',
    rewards: data.rewards || false,
    jobId: data.job_id || null,
    createdAt: data.created_at || '',
    updatedAt: data.updated_at || '',
    expiresAt: data.expires_at || null,
    meta: data.meta || null,
  };
}

export class Billing {
  constructor(private http: HTTPClient) {}

  /**
   * Get account balance
   */
  async balance(): Promise<Balance> {
    const data = await this.http.get('/api/balance');
    return balanceFromDict(data);
  }

  /**
   * Alias for balance() when callers want the richer shape explicitly.
   */
  async balanceDetails(): Promise<Balance> {
    return this.balance();
  }

  /**
   * List transactions
   */
  async transactions(limit: number = 50, page: number = 1): Promise<Transaction[]> {
    const data = await this.listTransactions({ page, pageSize: limit });
    return data.transactions;
  }

  /**
   * List transactions with filters and pagination metadata.
   */
  async listTransactions(options: ListTransactionsOptions = {}): Promise<TransactionListResponse> {
    const params: Record<string, string> = {
      page: String(options.page ?? 1),
      page_size: String(options.pageSize ?? 50),
    };
    if (options.transactionType) params.transaction_type = options.transactionType;
    if (options.status) params.status = options.status;
    if (options.jobId) params.job_id = options.jobId;

    const data = await this.http.get('/api/tx', params);
    const txList = data.transactions || [];
    return {
      transactions: txList.map(transactionFromDict),
      totalCount: data.total_count || txList.length,
    };
  }

  /**
   * Get a specific transaction
   */
  async getTransaction(transactionId: string): Promise<Transaction> {
    const data = await this.http.get(`/api/tx/${transactionId}`);
    return transactionFromDict(data);
  }

  /**
   * Create a Stripe Checkout session for balance top-ups.
   */
  async createTopUpCheckout(amount: number): Promise<TopUpCheckoutSession> {
    const data = await this.http.post('/api/stripe/top_up', { amount });
    return {
      sessionId: data.session_id || '',
      checkoutUrl: data.checkout_url || '',
    };
  }
}

File: rs-sdk/src/billing.rs
//! Account billing product API (`{product}/api/balance`, `{product}/api/tx`).
//!
//! Mirrors the Python SDK's `hypercli/billing.py`.

use secrecy::ExposeSecret;
use serde::Deserialize;

use crate::{HyperCliClient, HyperCliError};

/// Account balance (string amounts, `currency`/`decimals`-annotated).
#[derive(Clone, Debug, Deserialize, PartialEq)]
pub struct Balance {
    #[serde(default, alias = "total_balance")]
    pub total: String,
    #[serde(default, alias = "rewards_balance")]
    pub rewards: String,
    #[serde(default, alias = "balance")]
    pub paid: String,
    #[serde(default, alias = "available_balance")]
    pub available: String,
    #[serde(default = "default_currency")]
    pub currency: String,
    #[serde(default = "default_decimals")]
    pub decimals: u32,
}

fn default_currency() -> String {
    "USD".to_owned()
}

const fn default_decimals() -> u32 {
    2
}

/// One balance transaction.
#[derive(Clone, Debug, Deserialize, PartialEq)]
pub struct Transaction {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub user_id: String,
    #[serde(default)]
    pub amount: i64,
    #[serde(default)]
    pub amount_usd: f64,
    #[serde(default)]
    pub transaction_type: String,
    #[serde(default)]
    pub status: String,
    #[serde(default)]
    pub rewards: bool,
    #[serde(default)]
    pub job_id: Option<String>,
    #[serde(default)]
    pub created_at: String,
}

/// Blocking client for the product billing API, obtained through
/// [`HyperCliClient::billing`].
pub struct BillingClient<'a> {
    pub(crate) client: &'a HyperCliClient,
}

impl BillingClient<'_> {
    /// Get the account balance.
    pub fn balance(&self) -> Result<Balance, HyperCliError> {
        let url = self.client.product_endpoint("api/balance");
        self.client.send_json(
            "billing.balance",
            "GET",
            &url,
            None,
            self.client
                .http
                .get(&url)
                .bearer_auth(self.client.api_key.expose_secret()),
        )
    }

    /// List transactions (`limit` maps to the server's `page_size`).
    pub fn transactions(&self, limit: u32, page: u32) -> Result<Vec<Transaction>, HyperCliError> {
        let url = self.client.product_endpoint("api/tx");
        #[derive(Deserialize)]
        struct TransactionPage {
            #[serde(default)]
            transactions: Vec<Transaction>,
        }
        Ok(self
            .client
            .send_json::<TransactionPage>(
                "billing.transactions",
                "GET",
                &url,
                None,
                self.client
                    .http
                    .get(&url)
                    .bearer_auth(self.client.api_key.expose_secret())
                    .query(&[("page", page.to_string()), ("page_size", limit.to_string())]),
            )?
            .transactions)
    }

    /// Get one transaction.
    pub fn get_transaction(&self, transaction_id: &str) -> Result<Transaction, HyperCliError> {
        let url = self
            .client
            .product_endpoint(&format!("api/tx/{transaction_id}"));
        self.client.send_json(
            "billing.get_transaction",
            "GET",
            &url,
            None,
            self.client
                .http
                .get(&url)
                .bearer_auth(self.client.api_key.expose_secret()),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ClientConfig;
    use mockito::{Matcher, Server};
    use secrecy::SecretString;
    use serde_json::json;
    use url::Url;

    #[test]
    fn balance_and_transactions_round_trip() {
        let mut server = Server::new();
        let balance = server
            .mock("GET", "/api/balance")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(
                json!({
                    "total_balance": "10.00",
                    "rewards_balance": "1.00",
                    "balance": "9.00",
                    "available_balance": "8.00"
                })
                .to_string(),
            )
            .expect(1)
            .create();
        let tx = server
            .mock("GET", "/api/tx")
            .match_query(Matcher::AllOf(vec![
                Matcher::UrlEncoded("page".into(), "2".into()),
                Matcher::UrlEncoded("page_size".into(), "25".into()),
            ]))
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(
                json!({"transactions": [{"id": "tx-1", "amount": 500, "amount_usd": 5.0, "transaction_type": "charge", "status": "ok", "rewards": false}]})
                .to_string(),
            )
            .expect(1)
            .create();
        let tx_get = server
            .mock("GET", "/api/tx/tx-1")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(
                json!({"id": "tx-1", "amount": 500, "transaction_type": "charge", "status": "ok"})
                    .to_string(),
            )
            .expect(1)
            .create();
        let client = HyperCliClient::new(ClientConfig {
            api_base: Url::parse(&format!("{}/agents", server.url())).unwrap(),
            api_key: SecretString::from("test-credential"),
            trace_file: None,
            timeout: None,
        })
        .unwrap();

        let billing = client.billing();
        let fetched = billing.balance().unwrap();
        assert_eq!(fetched.total, "10.00");
        assert_eq!(fetched.paid, "9.00");
        assert_eq!(fetched.available, "8.00");
        assert_eq!(fetched.currency, "USD");
        assert_eq!(fetched.decimals, 2);

        let transactions = billing.transactions(25, 2).unwrap();
        assert_eq!(transactions.len(), 1);
        assert_eq!(transactions[0].id, "tx-1");
        assert_eq!(transactions[0].amount, 500);

        assert_eq!(billing.get_transaction("tx-1").unwrap().id, "tx-1");
        balance.assert();
        tx.assert();
        tx_get.assert();
    }
}
